lib/charms/mongodb/v0/upgrade_helpers.py

```diff
@@ -0,0 +1,464 @@
+#!/usr/bin/env python3
+"""Substrate agnostic manager for handling MongoDB in-place upgrades."""
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+import abc
+import logging
+import secrets
+import string
+from typing import List, Tuple
+
+from charms.mongodb.v0.mongo import MongoConfiguration
+from charms.mongodb.v1.mongodb import MongoDBConnection
+from charms.mongodb.v1.mongos import MongosConnection
+from ops.charm import CharmBase
+from ops.framework import Object
+from pymongo.errors import OperationFailure, PyMongoError, ServerSelectionTimeoutError
+from tenacity import Retrying, retry, stop_after_attempt, wait_fixed
+
+from config import Config
+
+# The unique Charmhub library identifier, never change it
+LIBID = "7f59c0d1c6f44d1993c57afa783e205c"
+
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+SHARD_NAME_INDEX = "_id"
+WRITE_KEY = "write_value"
+ROLLBACK_INSTRUCTIONS = "To rollback, `juju refresh` to the previous revision"
+
+
+# BEGIN: Useful classes
+class AbstractUpgrade(abc.ABC):
+    """In-place upgrades abstract class (typing)."""
+
+    pass
+
+
+# END: Useful classes
+
+
+# BEGIN: Exceptions
+class FailedToElectNewPrimaryError(Exception):
+    """Raised when a new primary isn't elected after stepping down."""
+
+
+class ClusterNotHealthyError(Exception):
+    """Raised when the cluster is not healthy."""
+
+
+class BalancerStillRunningError(Exception):
+    """Raised when the balancer is still running after stopping it."""
+
+
+# END: Exceptions
+
+
+class GenericMongoDBUpgrade(Object, metaclass=abc.ABCMeta):
```

Comment:
I thought it would be an issue because this could require that `Object` was also an abstract class but after a test, it doesn't seem so, so let's go for it.
Updated in my next commit